### PR TITLE
Improve Error Messages of Section 3

### PR DESCRIPTION
**Pull request by fsimonis**
Further information
* Displays the following when an interface was not found:
  ```
  (0) 11:10:36 [com::SocketCommunication]:885 in getIpAddress: ERROR: Cannot find network interface "loopy". Available interfaces are: lo enp0s31f6 br-3d4e767967a7 docker0 vethb6dcb86 Please check "network" attributes in your configuration file.
  ```
* Embeds python error messages in the preCICE error

Related to #698

### CODE REVIEW

**Comment by BenjaminRueth** [comment]
Is this PR ready for review? Then we should assign a reviewer.

File: src/xml/ConfigParser.cpp [hunk 1]
```diff
@@ -129,9 +129,7 @@ int ConfigParser::readXmlFile(std::string const &filePath)
   SAXHandler.serror         = OnStructuredErrorFunc;
 
   std::ifstream ifs(filePath);
-  if (not ifs) {
-    PRECICE_ERROR("File open error: " << filePath);
-  }
+  PRECICE_CHECK(ifs, "XML parser was unable to open configuration file \"" << filePath << '"');
```
Comment (BenjaminRueth) [inline src/xml/ConfigParser.cpp:132]: Why does this usually happen? Can we help the user?

File: src/com/MPISinglePortsCommunication.cpp [hunk 2]
```diff
@@ -79,12 +79,12 @@ void MPISinglePortsCommunication::acceptConnection(std::string const &acceptorNa
...
-                  "Requester communicator size has to be > 0!");
-    PRECICE_CHECK(requesterCommunicatorSize == peerCount,
-                  "Requester communicator sizes are inconsistent!");
+    PRECICE_ASSERT(requesterCommunicatorSize > 0,
+                  "Requester communicator size is " << requesterCommunicatorSize << " which is invalid.");
+    PRECICE_ASSERT(requesterCommunicatorSize == peerCount,
+                  "Current requester size from rank " << requesterRank << " is " << requesterCommunicatorSize<< " but should be " << peerCount);
     PRECICE_CHECK(_direct.count(requesterRank) == 0,
-                  "Duplicate request to connect by same rank (" << requesterRank << ")!");
+                  "Rank " << requesterRank << " has already been connected. Duplicate requests are not allowed.");
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/com/MPISinglePortsCommunication.cpp:87]: You are right, this should be an assertion :+1:

File: src/com/SocketCommunication.cpp [hunk 3]
```diff
@@ -99,7 +100,7 @@ void SocketCommunication::acceptConnection(std::string const &acceptorName,
       asio::read(*socket, asio::buffer(&requesterRank, sizeof(int)));
 
       PRECICE_CHECK(_sockets.count(requesterRank) == 0,
-                    "Duplicate request to connect by same rank (" << requesterRank << ")!");
+                    "Rank " << requesterRank << " has already been connected. Duplicate requests are not allowed.");
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/com/SocketCommunication.cpp:103]: This should also be an assertion

File: src/com/SocketCommunication.cpp [hunk 4]
```diff
@@ -375,7 +376,7 @@ void SocketCommunication::prepareEstablishment(std::string const &acceptorName,
   try {
     create_directories(dir);
   } catch (const boost::filesystem::filesystem_error &e) {
-    PRECICE_WARN("Creating directory for connection info failed with: " << e.what());
+    PRECICE_WARN("Creating directory for connection info failed with filesystem error: " << e.what());
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/com/SocketCommunication.cpp:379]: The `e.what()` contains the root of the error and the error description.
It looks a bit cryptic, but should be fine:
```
(0) 16:05:22 [com::SocketCommunication]:125 in acceptConnection: ERROR: Accepting connection at 127.0.0.1:47819 failed: boost::filesystem::create_directories: Not a directory: "./precice-run/SolverTwo-SolverOne"
```

File: src/io/ExportVTK.cpp [hunk 5]
```diff
@@ -42,7 +42,7 @@ void ExportVTK::doExport(
     fs::create_directories(outfile);
   outfile = outfile / fs::path(name + ".vtk");
   std::ofstream outstream(outfile.string(), std::ios::trunc);
-  PRECICE_CHECK(outstream, "Could not open file \"" << outfile << "\" for VTK export!");
+  PRECICE_CHECK(outstream, "VTK export failed to open destination file \"" << outfile << '"');
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/io/ExportVTK.cpp:45]: I don't think we can do anything without bloating the code too much.
Opening the file failed and the user has to check it.

File: src/io/ExportVTKXML.cpp [hunk 6]
```diff
@@ -151,7 +151,7 @@ void ExportVTKXML::writeSubFile(
   outfile = outfile / fs::path(name + "_r" + std::to_string(utils::MasterSlave::getRank()) + ".vtu");
   std::ofstream outSubFile(outfile.string(), std::ios::trunc);
 
-  PRECICE_CHECK(outSubFile, "Could not open slave file \"" << outfile.c_str() << "\" for VTKXML export!");
+  PRECICE_CHECK(outSubFile, "VTKXML export failed to open slave file \"" << outfile << '"');
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/io/ExportVTKXML.cpp:154]: https://github.com/precice/precice/pull/740#discussion_r454384516

File: src/io/ExportVTKXML.cpp [hunk 7]
```diff
@@ -83,7 +83,7 @@ void ExportVTKXML::writeMasterFile(
   outfile = outfile / fs::path(name + "_master.pvtu");
   std::ofstream outMasterFile(outfile.string(), std::ios::trunc);
 
-  PRECICE_CHECK(outMasterFile, "Could not open master file \"" << outfile.c_str() << "\" for VTKXML export!");
+  PRECICE_CHECK(outMasterFile, "VTKXML export failed to open master file \"" << outfile << '"');
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/io/ExportVTKXML.cpp:86]: https://github.com/precice/precice/pull/740#discussion_r454384516

File: src/io/TXTReader.cpp [hunk 8]
```diff
@@ -8,10 +8,8 @@ TXTReader::TXTReader(
     const std::string &filename)
     : _file()
 {
-  _file.open(filename.c_str());
-  if (not _file) {
-    PRECICE_ERROR("Could not open file \"" << filename << "\" for txt reading!");
-  }
+  _file.open(filename);
+  PRECICE_CHECK(_file, "TXT reader failed to open file \"" << filename << '"');
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/io/TXTReader.cpp:12]: https://github.com/precice/precice/pull/740#discussion_r454384516

File: src/io/TXTTableWriter.cpp [hunk 9]
```diff
@@ -14,11 +14,9 @@ TXTTableWriter::TXTTableWriter(
       _writeIterator(_data.end()),
       _outputStream()
 {
-  _outputStream.open(filename.c_str());
-  if (not _outputStream) {
-    PRECICE_ERROR("Could not open file \"" << filename
-                                           << "\" for writing txt table data!");
-  }
+  _outputStream.open(filename);
+  PRECICE_CHECK(_outputStream, "TXT table writer failed to open file \"" << filename << '"');
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/io/TXTTableWriter.cpp:18]: https://github.com/precice/precice/pull/740#discussion_r454384516

File: src/io/TXTWriter.cpp [hunk 10]
```diff
@@ -9,10 +9,9 @@ TXTWriter::TXTWriter(
     const std::string &filename)
     : _file()
 {
-  _file.open(filename.c_str());
-  if (not _file) {
-    PRECICE_ERROR("Could not open file \"" << filename << "\" for txt writing!");
-  }
+  _file.open(filename);
+  PRECICE_CHECK(_file, "TXT writer failed to open file \"" << filename << '"');
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/io/TXTWriter.cpp:13]: https://github.com/precice/precice/pull/740#discussion_r454384516

File: src/xml/ConfigParser.cpp (hunk 1 above)
Comment (fsimonis) [reply src/xml/ConfigParser.cpp:132]: https://github.com/precice/precice/pull/740#discussion_r454384516

**Pull request merged**